tests/test_eks/test_eks.py

```diff
@@ -0,0 +1,811 @@
+
+    result = client.list_clusters()[ResponseAttribute.CLUSTERS]
+
+    result.should.be.empty
+
+
+def test_list_clusters_returns_sorted_cluster_names(ClusterBuilder):
+    client, generated_test_data = ClusterBuilder(BatchCountSize.SMALL)
+    expected_result = sorted(generated_test_data.cluster_names)
+
+    result = client.list_clusters()[ResponseAttribute.CLUSTERS]
+
+    assert_result_matches_expected_list(result, expected_result, BatchCountSize.SMALL)
+
+
+def test_list_clusters_returns_default_max_results(ClusterBuilder):
+    client, generated_test_data = ClusterBuilder(BatchCountSize.LARGE)
+    expected_len = DEFAULT_MAX_RESULTS
+    expected_result = (sorted(generated_test_data.cluster_names))[:expected_len]
+
+    result = client.list_clusters()[ResponseAttribute.CLUSTERS]
+
+    assert_result_matches_expected_list(result, expected_result, expected_len)
+
+
+def test_list_clusters_returns_custom_max_results(ClusterBuilder):
+    client, generated_test_data = ClusterBuilder(BatchCountSize.MEDIUM)
+    expected_len = PageCount.LARGE
+    expected_result = (sorted(generated_test_data.cluster_names))[:expected_len]
+
+    result = client.list_clusters(maxResults=expected_len)[ResponseAttribute.CLUSTERS]
+
+    assert_result_matches_expected_list(result, expected_result, expected_len)
+
+
+def test_list_clusters_returns_second_page_results(ClusterBuilder):
+    client, generated_test_data = ClusterBuilder(BatchCountSize.MEDIUM)
+    page1_len = PageCount.LARGE
+    expected_len = BatchCountSize.MEDIUM - page1_len
+    expected_result = (sorted(generated_test_data.cluster_names))[page1_len:]
+    token = client.list_clusters(maxResults=page1_len)[ResponseAttribute.NEXT_TOKEN]
+
+    result = client.list_clusters(nextToken=token)[ResponseAttribute.CLUSTERS]
+
+    assert_result_matches_expected_list(result, expected_result, expected_len)
+
+
+def test_list_clusters_returns_custom_second_page_results(ClusterBuilder):
+    client, generated_test_data = ClusterBuilder(BatchCountSize.MEDIUM)
+    page1_len = PageCount.LARGE
+    expected_len = PageCount.SMALL
+    expected_result = (sorted(generated_test_data.cluster_names))[
+        page1_len : page1_len + expected_len
+    ]
+    token = client.list_clusters(maxResults=page1_len)[ResponseAttribute.NEXT_TOKEN]
+
+    result = client.list_clusters(maxResults=expected_len, nextToken=token)[
+        ResponseAttribute.CLUSTERS
+    ]
+
+    assert_result_matches_expected_list(result, expected_result, expected_len)
+
+
+def test_create_cluster_throws_exception_when_cluster_exists(ClusterBuilder):
+    client, generated_test_data = ClusterBuilder(BatchCountSize.SMALL)
+    expected_exception = ResourceInUseException
+    expected_msg = CLUSTER_EXISTS_MSG.format(
+        exception_name=expected_exception.TYPE,
+        method=MethodNames.CREATE_CLUSTER,
+        cluster_name=generated_test_data.existing_cluster_name,
+    )
+
+    with pytest.raises(expected_exception) as raised_exception:
```

Comment:
Apologies for not seeing this earlier - this happened to work before, because the exception handling wasn't implemented yet, but it should throw an error now that proper exception handling is in place.

Moto (like AWS) responds with a HTTP 4xx-error in this scenario, which boto3 then converts into a generic ClientError.
Any tests we run should therefore always verify that that specific ClientError is thrown.

Example:
```
from botocore.exceptions import ClientError
with pytest.raises(ClientError) as ex:
    ... something that fails
err = ex.value.response["Error"]
er["Message"].should.equal("..")
er["Code"].should.equal("..")
```

This is how it works against AWS, and is also the way to verify that we respond with a properly formatted HTTP 4xx-error that boto3 can understand.

When using this specific format, any test that passes the normal unit tests will also pass the ServerMode tests.

You might have to rewrite a few tests - I noticed there were a few more that follow the incorrect format.

Reply:
I see!   Alright, sounds good.  Between the commit you added above and this, I have a bit more work to do.  Apologies for all the revisions on this one; I'll get it there...

Reply:
This will add a **lot** of features, so some back-and-forth is expected :) 

The error handling is on us, really - there should be a standardized way of doing this. But with a 6+ year old 'organically grown' code base, and plenty of variations on what kind of output AWS/boto3 expects per service, it's not exactly easy to understand.

Reply:
Yeah, the fact that the output format of the AWS services vary so much doesn't make this any easier.

I'll get on adding the rest of those tests next week.  To make sure I understand what you have there and I'm not missing something: it looks like the error message format you were looking for happened to be just the exception name in this case, and not something like `"com.amazonaws.dynamodb.v20111205#ResourceNotFoundException"`, or is checking the header like that (test_server.py:L63) a shortcut/workaround?

> 
> Edit:
> The outstanding tasks would be:
> Ensure that the test_server tests pass, so that we know the error response is as expected
> Fix the negative tests per my comment below, so that all unit tests pass
> The ServerMode tests should then pass automagically as well

Thanks for this.

Reply:
> it looks like the error message format you were looking for happened to be just the exception name in this case, and not something like "com.amazonaws.dynamodb.v20111205#ResourceNotFoundException"

Yes, just the exception name. Dynamo is an exception in how it's named(of course..), because of the deprecated API version it refers to.

The header-check was added to ensure Moto's response is exactly the same as your Postman/curl-response. If AWS returns the exception name, so should we

Reply:
Alright. I think what I am seeing in your commit there is that on `exceptions.py:L13` the `self.description` should represent the expected curl result. In which case I should add the nodegroup name, addon name, and the short version of the message to that `dumps`, as verified in `test_server.py:L68`. Is that correct, or is the message being added somewhere behind the scenes that I am missing?

Reply:
Yes, that is correct